golang_src/cmd/worker.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"github.com/cerberus98/babymailgun"
+	"github.com/spf13/viper"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type WorkerConfig struct {
+	Host         string
+	Port         string
+	DatabaseName string
+	WorkerSleep  int
+	WorkerPool   int
+	MailHost     string
+	MailPort     string
+	AdminEmail   string
+}
+
+// Create a UUID4. Source implementation here: https://groups.google.com/forum/#!msg/golang-nuts/d0nF_k4dSx4/rPGgfXv6QCoJ
+func uuid() string {
+	b := make([]byte, 16)
+	rand.Read(b)
+	b[6] = (b[6] & 0x0f) | 0x40
+	b[8] = (b[8] & 0x3f) | 0x80
+	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
+}
+
+func processingLoop(cfg *WorkerConfig, wg *sync.WaitGroup, quit <-chan bool) {
+	// Fetch emails ready to be sent
+	// Try to send them
+	// Update the datastore
+	// Go back to sleep
+	log.Println("Starting processing loop...")
+	workerId := uuid()
+	defer wg.Done()
+loop:
+	for {
+		mongoClient := babymailgun.MongoClient{Host: cfg.Host, Port: cfg.Port, DatabaseName: cfg.DatabaseName}
+		select {
+		case <-quit:
+			log.Printf("Worker goroutine received quit")
+			break loop
+		default:
+			log.Println("Waking up and looking for emails to send")
+		}
+		email, err := mongoClient.FetchReadyEmail(workerId)
+		if err == nil {
+			log.Printf("Got email %s Worker ID: %s\n", email.ID, workerId)
+			// Try to send the email
+			mc := babymailgun.MailConfig{MailHost: cfg.MailHost, MailPort: cfg.MailPort, AdminEmail: cfg.AdminEmail}
```

Comment:
Do you expect to delivery emails only to one host that you have credentials available for? Or you assume that all emails go through a relay service?

Reply:
No, I'm going to prepopulate the db with a Server collection. It's just hardcoded to environment variables for now.